adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/BaseQualityRecalibration.scala

```diff
@@ -17,115 +17,335 @@
-        writer.println(readName + "\t" + String.valueOf(buf))
+  /**
+   * Observes the error covariates contained in a read.
+   *
+   * @param read The read to observe.
+   * @param recordGroups A record group dictionary containing the read group
+   *   this read is from.
+   * @param maskedSites The known SNP loci that this read covers.
+   * @return Returns an array of CovariateKeys that describe the per-base
+   *   error covariates seen in this read.
+   *
+   * @note This method was split out from the other observe method to make it
+   *   simpler to write unit tests for this package. Similarly, that is why
+   *   this method is package private, while the other is private.
+   */
+  private[recalibration] def observe(
+    read: AlignmentRecord,
+    recordGroups: RecordGroupDictionary,
+    maskedSites: Set[Long] = Set.empty): Array[CovariateKey] = ObservingRead.time {
+    val (toInclude, isMismatch) = ReadResidues.time {
+      computeResiduesToInclude(read, maskedSites)
+    }
+    ReadCovariates.time {
+      CovariateSpace(read,
+        toInclude,
+        isMismatch,
+        recordGroups)
     }
-    writer.close()
   }
-}
 
-private[read] object BaseQualityRecalibration {
+  /**
+   * Observes the error covariates contained in a read.
+   *
+   * @param read The read to observe.
+   * @param knownSnps A broadcast variable containing all known SNP loci to mask.
+   * @param recordGroups A record group dictionary containing the read group
+   *   this read is from.
+   * @return Returns an array of CovariateKeys that describe the per-base
+   *   error covariates seen in this read.
+   */
+  private def observe(
+    read: AlignmentRecord,
+    knownSnps: Broadcast[SnpTable],
+    recordGroups: RecordGroupDictionary): Array[CovariateKey] = ObservingRead.time {
+    val maskedSites = knownSnps.value
+      .maskedSites(ReferenceRegion.unstranded(read))
+    observe(read, recordGroups, maskedSites)
+  }
+
+  /**
+   * Runs base quality score recalibration.
+   *
+   * @param rdd The reads to recalibrate.
+   * @param knownSnps A broadcast table of known variation to mask during the
+   *   recalibration process.
+   * @param recordGroups The read groups that generated these reads.
+   * @param minAcceptableQuality The minimum quality score to attempt to
+   *   recalibrate.
+   * @param optStorageLevel An optional storage level to apply if caching the
+   *   output of the first stage of BQSR.
+   */
   def apply(
     rdd: RDD[AlignmentRecord],
     knownSnps: Broadcast[SnpTable],
-    observationDumpFile: Option[String] = None,
-    validationStringency: ValidationStringency = ValidationStringency.STRICT): RDD[AlignmentRecord] =
-    new BaseQualityRecalibration(cloy(rdd, validationStringency), knownSnps).result
+    recordGroups: RecordGroupDictionary,
+    minAcceptableQuality: Int,
+    optStorageLevel: Option[StorageLevel]): RDD[AlignmentRecord] = {
+    require(minAcceptableQuality >= 0 && minAcceptableQuality < 93,
```

Comment:
do we need a similar check above for `minAcceptableAsciiPhred`, or not worth it because it is private API?